modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,153 @@
+package org.tessellation.infrastructure.rewards
+
+import cats.data._
+import cats.effect.Async
+import cats.effect.std.Random
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.list._
+import cats.syntax.semigroup._
+
+import scala.collection.immutable.SortedSet
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Integral.Implicits._
+
+import org.tessellation.config.types.RewardsConfig
+import org.tessellation.dag.snapshot.epoch.EpochProgress
+import org.tessellation.domain.rewards._
+import org.tessellation.schema.ID.Id
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Amount
+import org.tessellation.schema.transaction.{RewardTransaction, TransactionAmount}
+import org.tessellation.security.SecurityProvider
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric._
+import io.estatico.newtype.ops.toCoercibleIdOps
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object Rewards {
+
+  def make[F[_]: Async: SecurityProvider](
+    config: RewardsConfig,
+    timeTriggerInterval: FiniteDuration,
+    softStaking: SoftStaking,
+    dtm: DTM,
+    stardust: StardustCollective
+  ): Rewards[F] =
+    new Rewards[F] {
+
+      implicit val logger = Slf4jLogger.getLogger[F]
+
+      def calculateRewards(
+        epochProgress: EpochProgress,
+        facilitators: NonEmptySet[Id]
+      ): F[SortedSet[RewardTransaction]] =
+        EitherT
+          .fromEither(getRewardsPerSnapshot(epochProgress))
+          .flatMap { snapshotReward =>
+            EitherT.liftF {
+              facilitators.toNonEmptyList
+                .traverse(_.toAddress)
+                .map(_.toNes)
+            }.flatMap(createSharedDistribution(epochProgress, snapshotReward, _))
+          }
+          // Note: Weights need to be validated
```

Comment:
Please compose the `apply*` methods in the following way:
1. Stardust collective (-10%)
2. DTM
3. Soft Staking
4. Distribution to facilitators

Reply:
Composition is from right to left

Reply:
True, I meant sequencing computations.

Reply:
To sequence methods in such way I need to change the output type of weight functions but as I said in the comment above - it will have some benefits.